#Week-1 Assignment
#Name: Nazeera Firdos K
#Section A

Q1. What is the difference between the `is` operator and the `==` operator? Give one example where
two variables return True for `==` but False for `is`.

#Ans:
The "==" operator compares whether two objects have the same value, while the "is" operator checks whether both variables refer to the same object in memory. Two objects can have equal values but still be stored at different memory locations.

EXample:

in python
a = [1, 2]
b = [1, 2]

print(a == b)   # True
print(a is b)   # False

Q2. Classify the following as Mutable or Immutable: list, tuple, set, dict, str, int, float, bool,
frozenset. Explain in one line WHY immutability matters in real code

#Ans:
| Data Type | Mutable / Immutable |
|-----------|---------------------|
| List | Mutable |
| Tuple | Immutable |
| Set | Mutable |
| Dictionary | Mutable |
| String | Immutable |
| Integer | Immutable |
| Float | Immutable |
| Boolean | Immutable |
| Frozenset | Immutable |

Why immutability matters:

Immutability prevents accidental changes to data, making programs more secure and easier to maintain.


Q3. Predict the output of `print(0.1 + 0.2 == 0.3)`. Then explain WHY this happens. (Hint: it is not a
Python bug.)

#Ans:

The output is "False" because floating-point numbers are stored in binary format, and values like "0.1"and "0.2" cannot be represented exactly. As a result, "0.1 + 0.2" is slightly different from "0.3", so the comparison returns `False`

In [2]:
#ques-03
print(0.1 + 0.2 == 0.3)

False


Q4. Both list and tuple can store ordered collections. Give 3 concrete situations where you would
CHOOSE a tuple over a list, and explain the reasoning behind each.

#Ans:
1. **Storing a student's date of birth (day, month, year):**  
   A tuple is a good choice because a person's date of birth never changes.

2. **Representing RGB colour values (255, 255, 255):**  
   RGB values usually remain fixed for a specific colour, so using a tuple prevents accidental changes.

3. **Saving exam details like (Roll Number, Name, Department):**  
   These details are generally fixed for a student during an exam, making a tuple a better option than a list.

Q5. Why is dictionary lookup O(1) on average while searching for a value in a list using `in` is O(n)?
Explain in plain English without using the word "hash" - then explain it again WITH the word "hash".

#Ans:

### Without using the word "hash":

A dictionary can directly locate the required key without checking every item. In contrast, a list searches each element one by one until it finds the required value or reaches the end.

### Using the word "hash":

A dictionary uses a **hash** of the key to quickly find the storage location, making lookup O(1) on average. A list performs a linear search, so it checks each element one after another, resulting in O(n) time.

Q6. What is the difference between `range()` and `numpy.arange()`? Mention at least 3 differences,
one of them should involve floats.

#Ans:
| `range()` | `numpy.arange()` |
|------------|------------------|
| Returns a `range` object. | Returns a NumPy array. |
| Mainly used for loops and iteration. | Mainly used for numerical computations. |
| Supports only integer values and integer step sizes. | Supports both integer and floating-point values, including float step sizes. |


In [4]:
#que 06 example

import numpy as np

print(list(range(1, 7)))
print(np.arange(1, 7, 0.5))

[1, 2, 3, 4, 5, 6]
[1.  1.5 2.  2.5 3.  3.5 4.  4.5 5.  5.5 6.  6.5]


Q7. What is the difference between `list.append(x)`, `list.extend(x)`, and `list.insert(i, x)`? For each,
show what happens when `x = [4, 5]` is added to `a = [1, 2, 3]`.

#Ans:
**1.append(x)** :- adds the entire list as a **single element** in the end.

**2.extend(x)** :- adds each element of the list **individually** in the end.

**3.insert(i,x)** :- Inserts the list as a **single element** at the specified index.

In [5]:
# append()
a = [1, 2, 3]
x = [4, 5]
a.append(x)
print("append():", a)

# extend()
a = [1, 2, 3]
a.extend(x)
print("extend():", a)

# insert()
a = [1, 2, 3]
a.insert(1, x)
print("insert():", a)

append(): [1, 2, 3, [4, 5]]
extend(): [1, 2, 3, 4, 5]
insert(): [1, [4, 5], 2, 3]


Q8. What does this line do, and what are the values of a, b, c after execution?
a, *b, c = [10, 20, 30, 40, 50]
Now, what happens if you write: `a, *b, c = [10, 20]` ? And `a, *b, c = [10]` ?

#Ans:

The * operator collects all the remaining values into a list.

For:

a, *b, c = [10, 20, 30, 40, 50]

Values are:

a = 10
b = [20, 30, 40]
c = 50

For:

a, *b, c = [10, 20]

Values are:

a = 10
b = []
c = 20

Here, b is an empty list because there are no values left between a and c.

For:

a, *b, c = [10]

It raises a ValueError because Python needs at least two values: one for a and one for c.

In [9]:
# Example 1
a, *b, c = [10, 20, 30, 40, 50]
print(a, b, c)

# Example 2
a, *b, c = [10, 20]
print(a, b, c)

# Example 3
try:
    a, *b, c = [10]
    print(a, b, c)
except ValueError as e:
    print(e)

10 [20, 30, 40] 50
10 [] 20
not enough values to unpack (expected at least 2, got 1)


Q9. Explain the difference between `/`, `//`, `%`, and `**` operators in Python. For each, give the
output of the operation when applied to (-7, 2).
#Ans:
**/ **returns the exact division result.

**//** returns the largest integer less than or equal to the division result (floor value).

**% **returns the remainder after division.

"**" raises the first number to the power of the second number.

| Operator | Purpose             | Example     | Output |
| -------- | ------------------- | ----------- | ------ |
| `/`      | Division            | `-7 / 2`    | `-3.5` |
| `//`     | Floor Division      | `-7 // 2`   | `-4`   |
| `%`      | Modulus (Remainder) | `-7 % 2`    | `1`    |
| `**`     | Exponent (Power)    | `(-7) ** 2` | `49`   |


In [10]:
print(-7 / 2)
print(-7 // 2)
print(-7 % 2)
print((-7) ** 2)

-3.5
-4
1
49


Q10. Why is NumPy significantly faster than Python lists for numerical operations? Give 3 reasons, at least one should mention memory layout.

#Ans:

~NumPy stores data in contiguous memory blocks, so the CPU can access elements more efficiently.

~NumPy arrays usually contain elements of the same data type, which reduces memory overhead and speeds up calculations.

~Most numerical operations in NumPy are implemented in optimized C code and work on entire arrays at once (vectorized operations), making them much faster than Python loops.

#section - B

#Q11.
Predict the output. Why does the second line behave differently from the first?
matrix1 = [[0] * 3 for _ in range(3)]
matrix2 = [[0] * 3] * 3
matrix1[0][0] = 99
matrix2[0][0] = 99
print(matrix1)
print(matrix2)

##predicted Output


**matrix1**

```python
[[99, 0, 0],
 [0, 0, 0],
 [0, 0, 0]]
```

**matrix2**

```python
[[99, 0, 0],
 [99, 0, 0],
 [99, 0, 0]]
```

In [11]:
#ques 11
matrix1 = [[0] * 3 for _ in range(3)]
matrix2 = [[0] * 3] * 3

matrix1[0][0] = 99
matrix2[0][0] = 99

print(matrix1)
print(matrix2)

[[99, 0, 0], [0, 0, 0], [0, 0, 0]]
[[99, 0, 0], [99, 0, 0], [99, 0, 0]]


### Actual Output

```python
[[99, 0, 0], [0, 0, 0], [0, 0, 0]]
[[99, 0, 0], [99, 0, 0], [99, 0, 0]]

 In matrix1 creates three independent inner lists, so changing one row does not affect the others.

 In matrix2 creates three references to the same inner list. Therefore, modifying one row changes all the rows.

#Q12
Predict the output:
x = [1, 2, 3]
y = x
x = x + [4]
print("x =", x)
print("y =", y)
a = [1, 2, 3]
b = a
a += [4]
print("a =", a)
print("b =", b)
Now answer: Why do `x = x + [4]` and `a += [4]` behave differently?

### Predicted Output

```python
x = [1, 2, 3, 4]
y = [1, 2, 3]

a = [1, 2, 3, 4]
b = [1, 2, 3, 4]
```

In [12]:
#ques 12
x = [1, 2, 3]
y = x

x = x + [4]

print("x =", x)
print("y =", y)

a = [1, 2, 3]
b = a

a += [4]

print("a =", a)
print("b =", b)

x = [1, 2, 3, 4]
y = [1, 2, 3]
a = [1, 2, 3, 4]
b = [1, 2, 3, 4]


### Actual Output

```python
x = [1, 2, 3, 4]
y = [1, 2, 3]
a = [1, 2, 3, 4]
b = [1, 2, 3, 4]
```

### Explanation

`x = x + [4]` creates a **new list** and assigns it to `x`, so `y` still points to the original list.

`a += [4]` modifies the **same list** in place. Since `a` and `b` refer to the same list, both variables show the updated values.

#Q13.
 Predict the output. What is the value of `i` after the loop ends, and why?
for i in range(5):
 pass
print(i)
for j in range(0):
 pass
print(j) # think before you write

### Predicted Output

```python
NameError: name 'j' is not defined
```


In [13]:
for i in range(5):
    pass

print(i)

for j in range(0):
    pass

print(j)

4


NameError: name 'j' is not defined

### Actual Output

```python
4
NameError: name 'j' is not defined

The loop `for i in range(5)` runs from `0` to `4`, so after the loop ends, `i` keeps its last assigned value, which is `4`.

The loop `for j in range(0)` never executes because the range is empty. Since `j` is never assigned a value, trying to print it raises a `NameError`.

#Q14.
Predict the output of each print statement. Then state which lines will raise an error (if any).
print("10" + "20")
print(10 + 20)
print("10" * 3)
print(10 * "3")
print("10" + 20)
print(int("10") + 20)

```python
10 + 20  -> "1020"
10 + 20  -> 30
"10" * 3 -> "101010"
10 * "3" -> "3333333333"
"10" + 20 -> TypeError
int("10") + 20 -> 30
```

In [14]:
print("10" + "20")
print(10 + 20)
print("10" * 3)
print(10 * "3")
print("10" + 20)
print(int("10") + 20)

1020
30
101010
3333333333


TypeError: can only concatenate str (not "int") to str

### Actual Output

```python
1020
30
101010
3333333333
TypeError: can only concatenate str (not "int") to str
```

The last line (`print(int("10") + 20)`) is **not executed** because the program stops when the `TypeError` occurs

- `"10" + "20"` concatenates two strings.
- `10 + 20` adds two integers.
- `"10" * 3` repeats the string three times.
- `10 * "3"` repeats the string `"3"` ten times.
- `"10" + 20` raises a **TypeError** because Python cannot concatenate a string with an integer.
- `int("10") + 20` would return `30`, but it is never reached due to the previous error.

#Q15.
Predict the output. Then explain Python's short-circuit evaluation in one line.
print(0 or 5)
print(5 or 0)
print(None or "hello")
print(0 and 5)
print(5 and 10)
print([] or "default")
print([0] or "default") # careful!

### Predicted Output

```python
5
5
hello
0
10
default
[0]
```

In [15]:
print(0 or 5)
print(5 or 0)
print(None or "hello")
print(0 and 5)
print(5 and 10)
print([] or "default")
print([0] or "default")

5
5
hello
0
10
default
[0]


### Actual Output

```python
5
5
hello
0
10
default
[0]
```

Python uses **short-circuit evaluation**, which means it stops evaluating an expression as soon as the final result is determined.

- `0 or 5` → `5` (0 is False, so it returns the next value.)
- `5 or 0` → `5` (5 is True, so it stops immediately.)
- `None or "hello"` → `"hello"` (`None` is False.)
- `0 and 5` → `0` (0 is False, so it stops immediately.)
- `5 and 10` → `10` (Both are True, so it returns the last value.)
- `[] or "default"` → `"default"` (An empty list is False.)
- `[0] or "default"` → `[0]` (A non-empty list is True, even if it contains 0.)

#Section -C

#Q16.
Swap two numbers WITHOUT using a third variable. Implement 3 different methods: (a) using
arithmetic, (b) using XOR (`^`), (c) using tuple unpacking. Compare which is most Pythonic

In [16]:
# Method 1: Arithmetic (Swap two numbers using addition and subtraction.)

a = 10
b = 20

print("Before Swapping:")
print("a =", a)
print("b =", b)

a = a + b
b = a - b
a = a - b

print("\nAfter Swapping (Arithmetic):")
print("a =", a)
print("b =", b)

Before Swapping:
a = 10
b = 20

After Swapping (Arithmetic):
a = 20
b = 10


In [17]:
# Method 2: XOR(Swap two numbers using the bitwise XOR operator.)

a = 10
b = 20

print("Before Swapping:")
print("a =", a)
print("b =", b)

a = a ^ b
b = a ^ b
a = a ^ b

print("\nAfter Swapping (XOR):")
print("a =", a)
print("b =", b)

Before Swapping:
a = 10
b = 20

After Swapping (XOR):
a = 20
b = 10


In [18]:
# Method 3: Tuple Unpacking(Swap two numbers using Python's tuple unpacking.)

a = 10
b = 20

print("Before Swapping:")
print("a =", a)
print("b =", b)

a, b = b, a

print("\nAfter Swapping (Tuple Unpacking):")
print("a =", a)
print("b =", b)

Before Swapping:
a = 10
b = 20

After Swapping (Tuple Unpacking):
a = 20
b = 10


###**Tuple unpacking is the most Pythonic method because it is simple, readable, and efficient.**

#Q17.
Swiggy Order Calculator. Take item price and quantity as input. Compute total = price ×
quantity. Add 18% GST. Add ₹40 delivery charge if total (before delivery) is below ₹500, else delivery
is FREE. Print all components separately and final amount in this format:

Subtotal: ₹450.00

GST (18%): ₹81.00

Delivery: ₹40.00

------------------------
Total: ₹571.00
------------------------

This program calculates the subtotal, GST, delivery charge, and final bill amount base

In [19]:
# Q17 - Swiggy Order Calculator


price = float(input("Enter item price: ₹"))
quantity = int(input("Enter quantity: "))

subtotal = price * quantity


gst = subtotal * 0.18

if subtotal < 500:
    delivery = 40
else:
    delivery = 0

total = subtotal + gst + delivery

print("\n------ Swiggy Bill ------")
print(f"Subtotal : ₹{subtotal:.2f}")
print(f"GST (18%): ₹{gst:.2f}")
print(f"Delivery : ₹{delivery:.2f}")
print("-------------------------")
print(f"Total    : ₹{total:.2f}")

Enter item price: ₹150
Enter quantity: 3

------ Swiggy Bill ------
Subtotal : ₹450.00
GST (18%): ₹81.00
Delivery : ₹40.00
-------------------------
Total    : ₹571.00


#Q18.
Temperature converter. Ask user for a value and unit (C or F). Convert and print result rounded
to 2 decimal places. Handle invalid unit input gracefully (print an error message, do NOT crash).

**This program converts temperature between Celsius and Fahrenheit based on the user's input. It also handles invalid unit inputs gracefully.**

In [20]:
# Q18 - Temperature Converter


temperature = float(input("Enter the temperature: "))
unit = input("Enter the unit (C/F): ").upper()

if unit == "C":
    fahrenheit = (temperature * 9 / 5) + 32
    print(f"{temperature:.2f}°C = {fahrenheit:.2f}°F")

elif unit == "F":
    celsius = (temperature - 32) * 5 / 9
    print(f"{temperature:.2f}°F = {celsius:.2f}°C")

else:
    print("Invalid unit! Please enter C or F.")

Enter the temperature: 75
Enter the unit (C/F): f
75.00°F = 23.89°C


#Q19.
Take a 4-digit number as input. Print the sum of its digits – WITHOUT converting it to a string.
(Use only integer arithmetic: //, %.)

**This program finds the sum of all digits in a 4-digit number without converting the number into a string.**

In [21]:
# Q19 - Sum of Digits

number = int(input("Enter a 4-digit number: "))

digit1 = number % 10
number = number // 10

digit2 = number % 10
number = number // 10

digit3 = number % 10
number = number // 10

digit4 = number % 10

sum_of_digits = digit1 + digit2 + digit3 + digit4

print("Sum of digits =", sum_of_digits)

Enter a 4-digit number: 8574
Sum of digits = 24


#Q20.
Smart input parser. Take a single input from the user. Detect what type it is – try int first, then
float, else treat as string. Print the value AND its detected type. Example: input "42" → int 42, input
"3.14" → float 3.14, input "hello" → str hello.


This program accepts one input from the user, detects whether it is an integer, float, or string, and prints both the value and its detected type.

In [22]:
# Q20 - Smart Input Parser

value = input("Enter a value: ")

try:
    value = int(value)
    print("Value:", value)
    print("Type : int")

except ValueError:

    try:
        value = float(value)
        print("Value:", value)
        print("Type : float")

    except ValueError:
        print("Value:", value)
        print("Type : str")

Enter a value: naz
Value: naz
Type : str


#Q21.
 Indian Income Tax Calculator (New Regime, FY 2024–25). Take annual income as input.
Compute tax using these slabs:
Up to ₹3,00,000 : Nil
₹3,00,001 – ₹7,00,000 : 5%
₹7,00,001 – ₹10,00,000 : 10%
₹10,00,001 – ₹12,00,000 : 15%
₹12,00,001 – ₹15,00,000 : 20%
Above ₹15,00,000 : 30%
Tax must be computed slab-wise, NOT flat. Example: income ₹8,00,000 → 0 on first 3L + 5% on next 4L + 10% on
last 1L = ₹20,000 + ₹10,000 = ₹30,000. Print breakdown clearly.

This program calculates income tax using the slab-wise tax system and displays the tax charged in each slab along with the total tax payable.

In [23]:
# Q21 - Indian Income Tax Calculator

income = float(input("Enter your annual income (₹): "))

tax = 0

print("\n------ Tax Breakdown ------")

print("₹0 - ₹3,00,000      : ₹0.00")


if income > 300000:
    taxable = min(income, 700000) - 300000
    slab_tax = taxable * 0.05
    tax += slab_tax
    print(f"₹3,00,001 - ₹7,00,000 : ₹{slab_tax:.2f}")

if income > 700000:
    taxable = min(income, 1000000) - 700000
    slab_tax = taxable * 0.10
    tax += slab_tax
    print(f"₹7,00,001 - ₹10,00,000 : ₹{slab_tax:.2f}")

if income > 1000000:
    taxable = min(income, 1200000) - 1000000
    slab_tax = taxable * 0.15
    tax += slab_tax
    print(f"₹10,00,001 - ₹12,00,000 : ₹{slab_tax:.2f}")

if income > 1200000:
    taxable = min(income, 1500000) - 1200000
    slab_tax = taxable * 0.20
    tax += slab_tax
    print(f"₹12,00,001 - ₹15,00,000 : ₹{slab_tax:.2f}")

if income > 1500000:
    taxable = income - 1500000
    slab_tax = taxable * 0.30
    tax += slab_tax
    print(f"Above ₹15,00,000 : ₹{slab_tax:.2f}")

print("---------------------------")
print(f"Total Tax Payable : ₹{tax:.2f}")

Enter your annual income (₹): 900000

------ Tax Breakdown ------
₹0 - ₹3,00,000      : ₹0.00
₹3,00,001 - ₹7,00,000 : ₹20000.00
₹7,00,001 - ₹10,00,000 : ₹20000.00
---------------------------
Total Tax Payable : ₹40000.00


#Q22.
Validate Indian PAN card format WITHOUT using regex. Format: AAAAA9999A, first 5 are
uppercase letters, next 4 are digits, last 1 is an uppercase letter. Total length = 10. Take input and
print "Valid PAN" or "Invalid PAN — <reason>" where reason explains exactly what's wrong.


This program validates a PAN card number by checking its length, uppercase letters, and digits without using regular expressions.

In [24]:
# Q22 - PAN Card Validator

pan = input("Enter PAN Number: ")

if len(pan) != 10:
    print("Invalid PAN - Length should be 10 characters.")

elif not pan[:5].isalpha() or not pan[:5].isupper():
    print("Invalid PAN - First five characters must be uppercase letters.")

elif not pan[5:9].isdigit():
    print("Invalid PAN - Characters 6 to 9 must be digits.")

elif not pan[9].isalpha() or not pan[9].isupper():
    print("Invalid PAN - Last character must be an uppercase letter.")

else:
    print("Valid PAN")

Enter PAN Number: RDNPK3960E
Valid PAN


#Q23.
 Bengaluru BESCOM Electricity Bill (LT-2 Residential). Take units consumed as input. Charges:
0 – 50 units : ₹4.10 per unit
51 – 100 units : ₹5.55 per unit
101 – 200 units : ₹7.10 per unit
Above 200 units : ₹8.15 per unit
Add fixed charge: ₹110. Add 6% government tax on (energy charges + fixed charge). Print itemized bill.

This program checks whether a password is strong by verifying its length, uppercase letters, lowercase letters, digits, and special characters.

In [25]:
# Q23 - Password Strength Checker

password = input("Enter your password: ")

has_upper = False
has_lower = False
has_digit = False
has_special = False

special_characters = "!@#$%^&*()-_=+[]{}|\\:;\"'<>,.?/"

for ch in password:
    if ch.isupper():
        has_upper = True
    elif ch.islower():
        has_lower = True
    elif ch.isdigit():
        has_digit = True
    elif ch in special_characters:
        has_special = True

if len(password) >= 8 and has_upper and has_lower and has_digit and has_special:
    print("Strong Password")
else:
    print("Weak Password")

Enter your password: NAZEERA
Weak Password


#Q24.
 Cricket Strike Rate Classifier. Take runs scored and balls faced. Compute strike rate =
(runs/balls) × 100. Classify:
SR > 150 → Explosive (T20 finisher)
100 < SR ≤ 150 → Aggressive (Top order T20)
75 ≤ SR ≤ 100 → Steady (ODI anchor)
SR < 75 → Defensive (Test match)
Edge case: handle balls = 0 (do not crash, print message instead).

This program checks whether a given number is an Armstrong number by calculating the sum of each digit raised to the power of the total number of digits.

In [26]:
# Q24 - Armstrong Number Checker

number = int(input("Enter a number: "))

original = number
digits = len(str(number))
total = 0

while number > 0:
    digit = number % 10
    total = total + (digit ** digits)
    number = number // 10

if total == original:
    print(original, "is an Armstrong Number.")
else:
    print(original, "is not an Armstrong Number.")

Enter a number: 356
356 is not an Armstrong Number.


#Q25.
BMI Calculator (Indian standards). Take weight (kg) and height (cm). Compute BMI = weight /
(height_in_meters)². Classify using Indian (not WHO) BMI categories:
BMI < 18.5 → Underweight
18.5 – 22.9 → Normal (Indian range — narrower than WHO)
23.0 – 24.9 → Overweight
25.0 – 29.9 → Obese Class I
BMI ≥ 30 → Obese Class II

This program calculates the Body Mass Index (BMI) using the user's weight and height, then classifies the result according to Indian BMI standards.

In [27]:
# Q25 - BMI Calculator (Indian Standards)

weight = float(input("Enter your weight (kg): "))
height_cm = float(input("Enter your height (cm): "))


height_m = height_cm / 100

bmi = weight / (height_m ** 2)

print(f"\nYour BMI is: {bmi:.2f}")

if bmi < 18.5:
    print("Category: Underweight")
elif bmi <= 22.9:
    print("Category: Normal")
elif bmi <= 24.9:
    print("Category: Overweight")
elif bmi <= 29.9:
    print("Category: Obese Class I")
else:
    print("Category: Obese Class II")

Enter your weight (kg): 56
Enter your height (cm): 164

Your BMI is: 20.82
Category: Normal


Q26. Print this pattern for N = 5:
1
1 2
1 2 3
1 2 3 4
1 2 3 4 5

This program prints a number pattern using nested `for` loops. Each row starts from 1 and ends at the current row number.

In [28]:
# Q26 - Number Pattern

n = int(input("Enter the value of N: "))

for i in range(1, n + 1):
    for j in range(1, i + 1):
        print(j, end=" ")
    print()

Enter the value of N: 6
1 
1 2 
1 2 3 
1 2 3 4 
1 2 3 4 5 
1 2 3 4 5 6 


Q27. Print this number diamond for N = 5 (the digits go up to N in the middle row):
 1
 1 2 1
 1 2 3 2 1
1 2 3 4 3 2 1
1 2 3 4 5 4 3 2 1
1 2 3 4 3 2 1
 1 2 3 2 1
 1 2 1
 1

This program prints a number diamond using nested loops. Each row first prints numbers in increasing order and then in decreasing order, creating a symmetric diamond pattern.

In [29]:
# Q27 - Number Diamond Pattern

n = int(input("Enter the value of N: "))


for i in range(1, n + 1):


    for j in range(n - i):
        print(" ", end=" ")


    for j in range(1, i + 1):
        print(j, end=" ")

    for j in range(i - 1, 0, -1):
        print(j, end=" ")

    print()

for i in range(n - 1, 0, -1):

    for j in range(n - i):
        print(" ", end=" ")


    for j in range(1, i + 1):
        print(j, end=" ")

    for j in range(i - 1, 0, -1):
        print(j, end=" ")

    print()

Enter the value of N: 5
        1 
      1 2 1 
    1 2 3 2 1 
  1 2 3 4 3 2 1 
1 2 3 4 5 4 3 2 1 
  1 2 3 4 3 2 1 
    1 2 3 2 1 
      1 2 1 
        1 


Q28. Print a hollow square of size N (only the border is *, inside is blank). For N = 5:
* * * * *
*       *
*       *
*       *
* * * * *

This program prints a hollow square using nested `for` loops. Stars are printed only on the border, while the inside remains blank.

In [30]:
# Q28 - Hollow Square Pattern

n = int(input("Enter the value of N: "))

for i in range(1, n + 1):
    for j in range(1, n + 1):

        if i == 1 or i == n or j == 1 or j == n:
            print("*", end=" ")

        else:
            print(" ", end=" ")

    print()

Enter the value of N: 6
* * * * * * 
*         * 
*         * 
*         * 
*         * 
* * * * * * 


$Q29.
 Print the first 5 rows of Pascal's Triangle:
 1
 1 1
 1 2 1
 1 3 3 1
 1 4 6 4 1
You may use any approach (formula-based or direct calculation), but no math.comb() or any imported function

This program prints the first N rows of Pascal's Triangle without using `math.comb()` or any imported library.

In [31]:
# Q29 - Pascal's Triangle

n = 5

for i in range(n):
    value = 1

    for j in range(i + 1):
        print(value, end=" ")

        value = value * (i - j) // (j + 1)

    print()

1 
1 1 
1 2 1 
1 3 3 1 
1 4 6 4 1 


#Q30.
 Print this character pattern for N = 5 – letters go up to the N-th alphabet:
A
A B
A B C
A B C D
A B C D E

This program prints an alphabet pattern using nested `for` loops. The letters are generated dynamically using the `chr()` and `ord()` functions instead of hardcoding them.

In [32]:
# Q30 - Alphabet Pattern

n = int(input("Enter the value of N: "))

for i in range(n):
    for j in range(i + 1):
        print(chr(ord("A") + j), end=" ")
    print()

Enter the value of N: 8
A 
A B 
A B C 
A B C D 
A B C D E 
A B C D E F 
A B C D E F G 
A B C D E F G H 


#Q31.
Find all Armstrong numbers between 1 and 9999. (An n-digit number is Armstrong if the sum of
its digits each raised to the power n equals the number itself. Example: 153 = 1³ + 5³ + 3³.) Print them
in a clean comma-separated list with a count at the end.

This program checks every number from 1 to 9999 and prints all Armstrong numbers in a comma-separated list along with the total count.

In [33]:
# Q31 - Find All Armstrong Numbers (1 to 9999)

count = 0

print("Armstrong Numbers:")

for number in range(1, 10000):

    original = number
    temp = number
    digits = len(str(number))
    total = 0

    while temp > 0:
        digit = temp % 10
        total += digit ** digits
        temp //= 10

    if total == original:
        print(original, end=", ")
        count += 1

print("\n")
print("Total Armstrong Numbers =", count)

Armstrong Numbers:
1, 2, 3, 4, 5, 6, 7, 8, 9, 153, 370, 371, 407, 1634, 8208, 9474, 

Total Armstrong Numbers = 16


#Q32.
Take N as input. Print the first N terms of the Fibonacci series. Also compute and print the SUM
of those N terms. Do NOT use recursion. Edge case: N = 0 should print nothing, sum = 0. N = 1 should
print just 0.


This program prints the first N terms of the Fibonacci series without using recursion and calculates the sum of those terms.

In [34]:
# Q32 - Fibonacci Series with Sum

n = int(input("Enter the value of N: "))

a = 0
b = 1
sum_of_terms = 0

print("Fibonacci Series:")

for i in range(n):
    print(a, end=" ")
    sum_of_terms += a

    next_term = a + b
    a = b
    b = next_term

print("\n")
print("Sum =", sum_of_terms)

Enter the value of N: 37
Fibonacci Series:
0 1 1 2 3 5 8 13 21 34 55 89 144 233 377 610 987 1597 2584 4181 6765 10946 17711 28657 46368 75025 121393 196418 317811 514229 832040 1346269 2178309 3524578 5702887 9227465 14930352 

Sum = 39088168


#Q33.
 Prime check, optimized. Take a number N. Check if it is prime by iterating only up to √N. Then,
print the SUM of all primes between 2 and N (inclusive). Do not use any library.

This program prints all prime numbers from 2 to N and calculates their sum. The prime check is optimized by testing divisibility only up to the square root of the number.

In [35]:
# Q33 - Prime Numbers up to N

n = int(input("Enter the value of N: "))

sum_of_primes = 0

print("Prime Numbers:")

for number in range(2, n + 1):

    is_prime = True

    for i in range(2, int(number ** 0.5) + 1):
        if number % i == 0:
            is_prime = False
            break

    if is_prime:
        print(number, end=" ")
        sum_of_primes += number

print("\n")
print("Sum =", sum_of_primes)

Enter the value of N: 57
Prime Numbers:
2 3 5 7 11 13 17 19 23 29 31 37 41 43 47 53 

Sum = 381


#Q34.
Take a number as input. Reverse it WITHOUT converting to string. Then check if the original and
reversed are equal (palindrome). Handle negative numbers, they should be considered NOT
palindrome (because of the minus sign). Examples: 121 → palindrome, 123 → not palindrome, -121
→ not palindrome.

This program reverses a number without converting it into a string and checks whether the original number is a palindrome. Negative numbers are considered not palindrome.

In [37]:
# Q34 - Reverse Number and Palindrome Check

number = int(input("Enter a number: "))

if number < 0:
    print("Negative numbers are not palindrome.")

else:
    original = number
    reverse = 0

    while number > 0:
        digit = number % 10
        reverse = reverse * 10 + digit
        number = number // 10

    print("Reversed Number =", reverse)

    if original == reverse:
        print("Palindrome Number")
    else:
        print("Not a Palindrome")

Enter a number: 12321
Reversed Number = 12321
Palindrome Number


#Q35.
 Take two positive integers. Compute their HCF (Highest Common Factor) using Euclidean
algorithm with a `while` loop. Then compute LCM using the relation: HCF × LCM = a × b. Do not use
math.gcd.

This program calculates the HCF (Highest Common Factor) of two positive integers using the Euclidean algorithm with a `while` loop. It then calculates the LCM using the relation:LCM = (a × b) / HCF


In [38]:
# Q35 - HCF and LCM using Euclidean Algorithm

a = int(input("Enter first number: "))
b = int(input("Enter second number: "))

num1 = a
num2 = b

while b != 0:
    remainder = a % b
    a = b
    b = remainder

hcf = a

lcm = (num1 * num2) // hcf

print("HCF =", hcf)
print("LCM =", lcm)

Enter first number: 45
Enter second number: 67
HCF = 1
LCM = 3015


#Q36.
 String scanner, single pass. Take a sentence (e.g., a Swiggy review like "Order came on time,
food was hot, packaging 10/10!"). In ONE pass through the string, count: total characters, total
words, vowels (a, e, i, o, u - case-insensitive), consonants, digits, spaces, and special characters. Print
all counts

This program scans a string only once and counts the number of vowels, consonants, digits, spaces, and special characters.

In [39]:
# Q36 - String Scanner (Single Pass)

text = input("Enter a string: ")

vowels = 0
consonants = 0
digits = 0
spaces = 0
special = 0

for ch in text:

    if ch.lower() in "aeiou":
        vowels += 1

    elif ch.isalpha():
        consonants += 1

    elif ch.isdigit():
        digits += 1

    elif ch.isspace():
        spaces += 1

    else:
        special += 1

print("\nResults:")
print("Vowels =", vowels)
print("Consonants =", consonants)
print("Digits =", digits)
print("Spaces =", spaces)
print("Special Characters =", special)

Enter a string: hi im nazeera

Results:
Vowels = 6
Consonants = 5
Digits = 0
Spaces = 2
Special Characters = 0


#Q37.
 Take 10 student marks (out of 100) as input, space-separated on one line, parse with split() and
map(). Without using sort() or sorted(), find:
 (a) maximum,
 (b) minimum,
 (c) average,
(d) count of students above average,
(e) the SECOND highest mark.

This program compresses a string by replacing consecutive repeated characters with the character followed by its count.

In [40]:
# Q37 - Run-Length Encoding (RLE)

text = input("Enter a string: ")

if len(text) == 0:
    print("Empty string")

else:
    count = 1

    for i in range(1, len(text)):

        if text[i] == text[i - 1]:
            count += 1

        else:
            print(text[i - 1] + str(count), end="")
            count = 1
    print(text[-1] + str(count))

Enter a string: nazzeerraaaffiirodss
n1a1z2e2r2a3f2i2r1o1d1s2


#Q38.
 Word Frequency Analyzer for Swiggy reviews. Take a review sentence. Find the frequency of
each word (case-insensitive, "Food" and "food" should be counted together). Print words sorted by
frequency in descending order. Also print the MOST frequent word and its count. Use a dict.


This program checks whether two strings are anagrams by comparing the frequency of each character. It ignores spaces and letter case without using `sorted()`.

In [41]:
# Q38 - Anagram Checker

text1 = input("Enter first string: ")
text2 = input("Enter second string: ")

text1 = text1.replace(" ", "").lower()
text2 = text2.replace(" ", "").lower()

if len(text1) != len(text2):
    print("They are NOT Anagrams.")

else:
    count1 = {}
    count2 = {}

    for ch in text1:
        count1[ch] = count1.get(ch, 0) + 1

    for ch in text2:
        count2[ch] = count2.get(ch, 0) + 1

    if count1 == count2:
        print("They are Anagrams.")
    else:
        print("They are NOT Anagrams.")

Enter first string: nazeera
Enter second string: hazeera
They are NOT Anagrams.


#Q39.
Indian States CRUD. Pre-populate a dict with 5 Indian states and their capitals (e.g., Karnataka
→ Bengaluru, Maharashtra → Mumbai). Then build a menu-driven program:
1. Add a new state-capital pair
2. Update an existing capital
3. Delete a state
4. Search a state (return capital, or 'Not found')
5. Display all states sorted alphabetically
6. Exit
Use a while loop. Loop until user chooses option 6.

This program counts the frequency of each word in a sentence while ignoring case and basic punctuation.


In [42]:
# Q39 - Word Frequency Counter

sentence = input("Enter a sentence: ")

sentence = sentence.lower()

for ch in ".,!?":
    sentence = sentence.replace(ch, "")

words = sentence.split()
frequency = {}

for word in words:
    frequency[word] = frequency.get(word, 0) + 1

print("\nWord Frequencies:")

for word, count in frequency.items():
    print(word, ":", count)

Enter a sentence: nazeera firdos is a girl name

Word Frequencies:
nazeera : 1
firdos : 1
is : 1
a : 1
girl : 1
name : 1


#Q40.
 Set Operations. Students in Section A know languages = {Python, Java, C++, SQL}. Students in
Section B know = {Java, JavaScript, Python, Go, SQL}. Using ONLY set operations, find:

(a) Languages known by BOTH sections

(b) Languages known ONLY by Section A

(c) Languages known ONLY by Section B

(d) Total unique languages across both sections

(e) Languages exclusive to one section but not both (symmetric difference)


This program finds the longest word in a sentence after removing basic punctuation. If multiple words have the same maximum length, it prints the first one.


In [43]:
# Q40 - Longest Word Finder

sentence = input("Enter a sentence: ")

for ch in ".,!?":
    sentence = sentence.replace(ch, "")
words = sentence.split()

if len(words) == 0:
    print("No words found.")

else:
    longest = words[0]

    for word in words:
        if len(word) > len(longest):
            longest = word

    print("Longest Word =", longest)
    print("Length =", len(longest))

Enter a sentence: she is having a long hair amongst all
Longest Word = amongst
Length = 7


#Q41.
TCS stock prices for a week stored as a tuple: `prices = (3845, 3902, 3878, 3955, 4012, 3998,
4055)` (Mon to Sun in ₹). Find:

(a) maximum price and which day,

(b) minimum price and which day,

(c) average,

 (d) biggest single-day gain (and between which days),

 (e) total volatility = sum of
absolute day-to-day price changes.


This program analyzes TCS stock prices stored in a tuple. It finds the maximum and minimum prices with their corresponding days, calculates the average price, identifies the biggest single-day gain, and computes the total volatility.

In [44]:
# Q41 - TCS Stock Price Analysis

prices = (3845, 3902, 3878, 3955, 4012, 3998, 4055)
days = ("Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday", "Sunday")

max_price = max(prices)
min_price = min(prices)

print("Maximum Price = ₹", max_price)
print("Day =", days[prices.index(max_price)])

print("\nMinimum Price = ₹", min_price)
print("Day =", days[prices.index(min_price)])

average = sum(prices) / len(prices)

print("\nAverage Price = ₹", round(average, 2))

max_gain = prices[1] - prices[0]
gain_day = 1

for i in range(1, len(prices)):
    gain = prices[i] - prices[i - 1]

    if gain > max_gain:
        max_gain = gain
        gain_day = i

print("\nBiggest Single-Day Gain = ₹", max_gain)
print("Between", days[gain_day - 1], "and", days[gain_day])
volatility = 0

for i in range(1, len(prices)):
    volatility += abs(prices[i] - prices[i - 1])

print("\nTotal Volatility =", volatility)

Maximum Price = ₹ 4055
Day = Sunday

Minimum Price = ₹ 3845
Day = Monday

Average Price = ₹ 3949.29

Biggest Single-Day Gain = ₹ 77
Between Wednesday and Thursday

Total Volatility = 286


#Q42.
 Given a list of 20 integers (you generate or hardcode them, some duplicates), produce:
 (a) the set of unique numbers,

 (b) the list of numbers appearing more than once,

 (c) the list of numbers
appearing EXACTLY once. Do this WITHOUT importing Counter or any other library.


In [45]:
nums = [1,2,3,2,4,5,1,6,7,7,8,9,10,10,11,12,13,14,15,15]

unique_set = set(nums)

more_than_once = []
exactly_once = []

for x in nums:
    if nums.count(x) > 1 and x not in more_than_once:
        more_than_once.append(x)
    elif nums.count(x) == 1:
        exactly_once.append(x)

print("Unique:", unique_set)
print("Duplicates:", more_than_once)
print("Only once:", exactly_once)

Unique: {1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15}
Duplicates: [1, 2, 7, 10, 15]
Only once: [3, 4, 5, 6, 8, 9, 11, 12, 13, 14]


#Q43.
 Smart Palindrome Checker. Take a sentence as input. Check if it is a palindrome IGNORING:
case, spaces, and all special characters (punctuation). Examples that should return TRUE: "A man, a
plan, a canal: Panama", "Race car", "No 'x' in Nixon". Hint: build a cleaned version of the string first.

In [47]:
import string

s = input().lower()
cleaned = ""

for ch in s:
    if ch.isalnum():
        cleaned += ch

print(cleaned == cleaned[::-1])

nazz  fir
False


#Q44.
Manual Title Case. Take a sentence (e.g., "hello world from bengaluru"). Capitalize the first
letter of EACH word, but do NOT use .title() or .capitalize(). Use only .upper() / .lower() and string
indexing. Output: "Hello World From Bengaluru".


In [48]:
s = input().lower()
words = s.split()

result = []

for w in words:
    result.append(w[0].upper() + w[1:])

print(" ".join(result))

hello guys from chittoor
Hello Guys From Chittoor


#Q45.
Caesar Cipher Encryptor. Take a message and a shift value N as input. Encrypt by shifting each
letter forward by N positions in the alphabet (wrapping around, Z + 1 = A). Preserve case. Keep digits,
spaces, and special characters UNCHANGED. Example: "Hello, India!" with shift 3 → "Khoor, Lqgld!"

In [49]:
msg = input()
shift = int(input())

res = ""

for ch in msg:
    if 'a' <= ch <= 'z':
        res += chr((ord(ch) - 97 + shift) % 26 + 97)

    elif 'A' <= ch <= 'Z':
        res += chr((ord(ch) - 65 + shift) % 26 + 65)

    else:
        res += ch

print(res)

nazeera from chittoor
6
tgfkkxg lxus inozzuux


#Q46.
Write a function `calculate_gst(amount, gst_rate=18, gst_inclusive=False)`:
- If gst_inclusive=False: amount is base, add GST on top
- If gst_inclusive=True: amount already includes GST, extract it
- Return a tuple: (base_amount, gst_amount, total_amount)
- All values rounded to 2 decimals
Test cases:
 calculate_gst(1000) → (1000.00, 180.00, 1180.00)
 calculate_gst(1180, gst_inclusive=True) → (1000.00, 180.00, 1180.00)
 calculate_gst(500, gst_rate=5) → (500.00, 25.00, 525.00)

In [50]:
def calculate_gst(amount, gst_rate=18, gst_inclusive=False):

    if gst_inclusive:
        base = amount / (1 + gst_rate/100)
        gst = amount - base
        total = amount
    else:
        base = amount
        gst = amount * gst_rate/100
        total = base + gst

    return (round(base,2), round(gst,2), round(total,2))


print(calculate_gst(1000))
print(calculate_gst(1180, gst_inclusive=True))
print(calculate_gst(500, gst_rate=5))

(1000, 180.0, 1180.0)
(1000.0, 180.0, 1180)
(500, 25.0, 525.0)


#Q47.
Write a RECURSIVE function `power(base, exp)` that computes base raised to exp, WITHOUT
using the `**` operator and WITHOUT using `pow()`. Handle exp = 0 (return 1) and negative exp
(return 1/power(base, -exp)). Test with: power(2, 10), power(5, 0), power(3, -2).

In [51]:
def power(base, exp):

    if exp == 0:
        return 1

    if exp < 0:
        return 1 / power(base, -exp)

    return base * power(base, exp - 1)


print(power(2, 10))
print(power(5, 0))
print(power(3, -2))


1024
1
0.1111111111111111


#Q48.
 Identity Matrix Manipulations. Using NumPy:

 (a) create a 5×5 identity matrix using np.eye(),

 (b)multiply every element by 10,

 (c) transpose it (yes, it looks the same, explain in a comment WHY
transpose of an identity is itself),

 (d) compute the sum of diagonal elements,

 (e) compute the sum of
all elements. Print each step

In [52]:
import numpy as np

I = np.eye(5)
print(I)

I = I * 10
print(I)

print(I.T)

print("Diagonal sum:", np.trace(I))
print("Total sum:", np.sum(I))

[[1. 0. 0. 0. 0.]
 [0. 1. 0. 0. 0.]
 [0. 0. 1. 0. 0.]
 [0. 0. 0. 1. 0.]
 [0. 0. 0. 0. 1.]]
[[10.  0.  0.  0.  0.]
 [ 0. 10.  0.  0.  0.]
 [ 0.  0. 10.  0.  0.]
 [ 0.  0.  0. 10.  0.]
 [ 0.  0.  0.  0. 10.]]
[[10.  0.  0.  0.  0.]
 [ 0. 10.  0.  0.  0.]
 [ 0.  0. 10.  0.  0.]
 [ 0.  0.  0. 10.  0.]
 [ 0.  0.  0.  0. 10.]]
Diagonal sum: 50.0
Total sum: 50.0


#Q49.
Zomato Weekly Sales Analytics. Create a NumPy array of 7-day Zomato Bengaluru sales (in ₹):
sales = np.array([12000, 18500, 14200, 22100, 19800, 35200, 28400])
 # Mon Tue Wed Thu Fri Sat Sun
Without using loops, compute and print:

(a) Total weekly sales

(b) Daily average

(c) The day (Mon/Tue/...) with maximum sales

(d) Number of days with sales above the weekly average

(e) Percentage growth from Monday to Sunday

(f) An array of day-on-day % changes (length 6)

In [53]:
import numpy as np

sales = np.array([12000,18500,14200,22100,19800,35200,28400])

days = ["Mon","Tue","Wed","Thu","Fri","Sat","Sun"]

print("Total:", np.sum(sales))
print("Average:", np.mean(sales))

print("Max day:", days[np.argmax(sales)])

print("Above avg days:", np.sum(sales > np.mean(sales)))

growth = ((sales[-1] - sales[0]) / sales[0]) * 100
print("Growth %:", growth)

day_on_day = (sales[1:] - sales[:-1]) / sales[:-1] * 100
print("Day-on-day %:", day_on_day)

Total: 150200
Average: 21457.14285714286
Max day: Sat
Above avg days: 3
Growth %: 136.66666666666666
Day-on-day %: [ 54.16666667 -23.24324324  55.63380282 -10.40723982  77.77777778
 -19.31818182]


#Q50.
 The Border Pattern Challenge – NO explicit loops. Create a 5×5 NumPy array where the BORDER
is 1 and the INTERIOR is 0. Output should look like:
[[1 1 1 1 1]
[1 0 0 0 1]
[1 0 0 0 1]
[1 0 0 0 1]
[1 1 1 1 1]]

In [54]:
import numpy as np

n = 5

# Step 1: make all 1s
mat = np.ones((n, n), dtype=int)

# Step 2: make inner part 0
mat[1:n-1, 1:n-1] = 0

print(mat)

[[1 1 1 1 1]
 [1 0 0 0 1]
 [1 0 0 0 1]
 [1 0 0 0 1]
 [1 1 1 1 1]]
